vignettes/hdf5.Rmd

```diff
@@ -0,0 +1,159 @@
+  - SIB Swiss Institute of Bioinformatics
+  email: charlottesoneson@gmail.com
+- name: Aaron Lun
+  affiliation: 
+  - &id2 Cancer Research UK Cambridge Institute, University of Cambridge
+  email: infinite.monkeys.with.keyboards@gmail.com 
+date: "`r BiocStyle::doc_date()`"
+package: "`r BiocStyle::pkg_ver('iSEE')`"
+output: 
+  BiocStyle::html_document:
+    toc_float: true
+vignette: >
+  %\VignetteIndexEntry{6. Using iSEE with HDF5}
+  %\VignetteEncoding{UTF-8}  
+  %\VignettePackage{iSEE}
+  %\VignetteKeywords{GeneExpression, RNASeq, Sequencing, Visualization, QualityControl, GUI}
+  %\VignetteEngine{knitr::rmarkdown}
+editor_options: 
+  chunk_output_type: console
+bibliography: iSEE.bib
+---
+
+**Compiled date**: `r Sys.Date()`
+
+**Last edited**: 2018-03-08
+
+**License**: `r packageDescription("iSEE")[["License"]]`
+
+```{r setup, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>",
+  error = FALSE,
+  warning = FALSE,
+  message = FALSE,
+    crop = NULL
+)
+stopifnot(requireNamespace("htmltools"))
+htmltools::tagList(rmarkdown::html_dependency_font_awesome())
+sce <- readRDS("sce.rds")
+```
+
+```{r, eval=!exists("SCREENSHOT"), include=FALSE}
+SCREENSHOT <- function(x, ...) knitr::include_graphics(x)
+```
+
+# Wrapping matrices in HDF5 files
+
+Let us make a copy of the originak `sce` object to work with; we will be comparing the memory footprint of the object using an HDF5 backend.
+
+```{r}
+sce_h5 <- sce
+```
+
+Many `SummarizedExperiment` objects store assay matrices as regular `matrix` objects.
+
+```{r}
+library(SingleCellExperiment)
+class(assay(sce_h5, "logcounts"))
+```
+
+This type of base `matrix` is fully loaded in memory, which can become prohibitive for very large datasets.
+
+However, the HDF5 file format provides an alternative to work with large datasets.
+
+> HDF5 is a unique technology suite that makes possible the management of extremely large and complex data collections.
+> (Source: https://support.hdfgroup.org/HDF5/whatishdf5.html)
+
+Array-like objects such as matrices can be wrapped in an `HDF5Matrix` object as follows.
+
+```{r}
+library(HDF5Array)
+assay(sce_h5, "tophat_counts") <- writeHDF5Array(assay(sce_h5, "tophat_counts"))
+class(assay(sce_h5, "tophat_counts"))
+assay(sce_h5, "logcounts") <- writeHDF5Array(assay(sce_h5, "logcounts"))
+class(assay(sce_h5, "logcounts"))
+```
+
+Comparing the memory footprint of the backed up original object `sce` and the HDF5-backed `sce_h5` object reveals the smallermemory footprint of HDF5-backed objects.
+
+```{r}
+format(object.size(sce), units = "Mb")
+format(object.size(sce_h5), units = "Mb")
+```
+
+```{r, results='asis', echo=FALSE}
+cat(sprintf(
+  "In this case, using an HDF5 backend reduced the memory footprint %.1f-fold.",
+  object.size(sce) / object.size(sce_h5)))
+```
+
+# iSEE supports HDF5-backed assays
+
+We can then used the `SingleCellExperiment` object as usual.
```

Comment:
Mention that this works for any (i) DelayedArray backend (e.g., `ResidualMatrix`, `LowRankMatrix` from the **BiocSingular** package; `RleMatrix` objects; SQL-backed stores, though I don't know if these exist yet) and (ii) any matrix-like object in general that supports `dim()` and `[`.